### DISCUSSION TITLE
Inquiry Regarding Crazyflie 2.1 Webots Simulation

### DISCUSSION THREAD

**Discussion by mrrranjbar** [bitcraze/discussions#1625] [Q&A]
I am working on motion planning for the Crazyflie 2.1 using deep reinforcement learning. The trained model performs well in the Webots simulation, where I implement high-level velocity control (v_x, v_y, v_z, and yaw as desired) and rely on the existing PID controllers in both the simulation and the real drone. However, I observe a notable discrepancy between the simulation and real-world experiments. Specifically, the PID gains in the simulation seem to cause the drone to behave more “cautiously” or “sluggishly” compared to its real counterpart.

PID gains in Webots = {"kp_att_y": 1, "kd_att_y": 0.5, "kp_att_rp": 0.5, "kd_att_rp": 0.1, "kp_vel_xy": 2, "kd_vel_xy": 0.5, "kp_z": 10, "ki_z": 5, "kd_z": 5}

PID gains in real Crazyflie 2.1 = {"kp_att_y": 6, "kd_att_y": 0.35, "kp_att_rp": 6, "kd_att_rp": 0, "kp_vel_xy": 25, "kd_vel_xy": 0, "kp_z": 2, "ki_z": 0.5, "kd_z": 0} based on ["platform_defaults_cf21bl.h"](src/platform/interface/platform_defaults_cf21bl.h)

Do you have any insights or recommendations on how to address this issue? My goal is to create a simulation environment that accurately mirrors the real-world dynamics of Crazyflie 2.1.

Thank you very much for your time and assistance.

**Comment by knmcguire** [comment]
Hi!

So the webots standard example for the Crazyflie flies on a different PID controller than the actual Crazyflie. Are you using the regular example or the pythonbinding version? 
 Also the webots air dynamics models is quite simple though as well. The wall following example is pretty one-on-one in terms of behavior but for deep reinforcement learning I'm not sure if that is the most suitable.

Have you considered looking at Crazysim or Sim_CF2? They have created a full software in the loop version of the crazyflie firmware and use gazebo as the similator. 
Mind that our inhouse simulation support is limited and experimental, but there are quite some external projects out there that you can look at as well. Here is [a page](https://www.bitcraze.io/development/external-projects/#crazyflie-simulators) where we link to those on our website
